### ISSUE TITLE
[Bug] TROPOMI test fails on my MacBook, but passes in CI and on my iMac

### ISSUE THREAD

**Issue by srherbener** [JCSDA-internal/ioda-converters#616]
## Current behavior (describe the bug)

The ctest `test_iodaconv_tropomi_no2` passes on my iMac and in the CI testing, but fails on my MacBook with the following error message:

```
469/495 Testing: test_iodaconv_tropomi_no2
469/495 Test: test_iodaconv_tropomi_no2
Command: "/Users/steveherbener/projects/CONVERTER_V2/ioda-bundle/build/bin/iodaconv_comp.sh" "netcdf" "/Users/steveherbener/projects/CONVERTER_V2/ioda-bundle/build/bin/tropomi_no2_nc2ioda.py
                          -i testinput/tropomi_no2.nc
                          -o testrun/tropomi_no2.nc" "tropomi_no2.nc" "0.0"
Directory: /Users/steveherbener/projects/CONVERTER_V2/ioda-bundle/build/iodaconv/test
"test_iodaconv_tropomi_no2" start time: Sep 09 15:40 MDT
Output:
----------------------------------------------------------
Assertion failed: (H5VL_def_conn_s.connector_id == (-1)), function H5VL__set_def_conn, file H5VLint.c, line 401.
/Users/steveherbener/projects/CONVERTER_V2/ioda-bundle/build/bin/iodaconv_comp.sh: line 20: 30683 Abort trap: 6           $cmd
<end of output>
Test time =   0.53 sec
----------------------------------------------------------
Test Failed.
"test_iodaconv_tropomi_no2" end time: Sep 09 15:40 MDT
"test_iodaconv_tropomi_no2" time elapsed: 00:00:00
----------------------------------------------------------
```

## To Reproduce

It's not clear how to reproduce this yet. Here are some differences between my iMac and MacBook which are potential culprits.

My iMac configuration is:

-  Big Sur 11.5.1
-  Clang 12.0.5
-  GNU Fortran 11.1.0

My MacBook configuration is:

- Big Sur 11.5.2
- Clang 12.0.5
- GNU Fortran 11.2.0

## Expected behavior

All ioda-converter ctests pass.

**Comment by jamiebresch** [comment]
I think OS and compiler versions have little to do with this issue.

I managed to run stand-alone tropomi_no2_nc2ioda.py successfully.
`setenv $PYTHONPATH /Users/hclin/code/jedi/ioda-bundle/build0909/lib/python3.9`

With brew installed python and specify python command (it works):
```
/usr/local/bin/python ../../bin/tropomi_no2_nc2ioda.py -i testinput/tropomi_no2.nc -o testrun/out.nc
```

Without specifying the python command (strange error as in ctest):
```
../../bin/tropomi_no2_nc2ioda.py -i testinput/tropomi_no2.nc -o testrun/out.nc
Assertion failed: (H5VL_def_conn_s.connector_id == (-1)), function H5VL__set_def_conn, file H5VLint.c, line 401.
Abort
```
However, adding /usr/local/bin/python to the test command ARGS in iodaconv/test/CMakeLists.txt didn't fix the ctest error.

With conda python:
```
/opt/anaconda3/bin/python ../../bin/tropomi_no2_nc2ioda.py -i testinput/tropomi_no2.nc -o testrun/out.nc
Traceback (most recent call last):
  File "../../bin/tropomi_no2_nc2ioda.py", line 23, in <module>
    import ioda_conv_engines as iconv
  File "/Users/hclin/code/jedi/ioda-bundle/build0909/lib/pyiodaconv/ioda_conv_engines.py", line 2, in <module>
    import ioda
  File "/Users/hclin/code/jedi/ioda-bundle/build0909/lib/python3.9/ioda/__init__.py", line 1, in <module>
    from .ioda import *
  File "/Users/hclin/code/jedi/ioda-bundle/build0909/lib/python3.9/ioda/ioda.py", line 1, in <module>
    import ioda._ioda_python as _iodapy
ModuleNotFoundError: No module named 'ioda._ioda_python'
```

**Comment by srherbener** [comment]
@jamiebresch thank you for your investigation. I get the same behavior on my MacBook, and I was able to find the cause of this issue by looking at differences in the environment when running the script the two ways you mentioned.

The difference is related to whether /usr/bin/env is being used. A few years ago, Apple introduce a feature called System Integrity Protection (https://support.apple.com/en-us/HT204899) to help fight malware and viruses. Unfortunately a side effect of this was to disallow "dangerous" environment variables (LD_LIBRARY_PATH, eg) to be set for executables in /usr/bin and other places. So when using /usr/bin/env, the LD_LIBRARY_PATH and DYLD_LIBRARY_PATH variables are quietly unset before running python3. This does not happen with python3 (as long as it is not installed in /usr/bin).

The unsetting of the library path variables causes the hdf5 library to be picked up from another place than jedi stack, likely the python installation itself due to h5py or netCDF4 modules being installed. This library is out of sync with what the ioda python bindings were compiled with, thus the cryptic error.

You can disable the System Integrity Protection on you Mac using these instructions: https://developer.apple.com/documentation/security/disabling_and_enabling_system_integrity_protection

I disabled this on my MacBook and the tropomi test is passing now.

Could you try this on your MacBook and see if it fixes the problem? Thanks!

**Comment by jamiebresch** [comment]
Impressive work! Great explanation and information. It is good enough for me to know what is happening. Using `/usr/bin/env python3` is supposedly for portability. Now I know the behavior on my MacBook, my own preferred solution is to specify explicit python path. Thanks @srherbener for figuring things out!

**Comment by srherbener** [comment]
We have changed ioda converter tests to run the scripts via `python3 path_to_script_being_tested args` which also works to preserve LD_LIBRARY_PATH, DYLD_LIBRARY_PATH. This is so since /usr/bin/env is not called to find the interpreter.

This change was done in this PR: https://github.com/JCSDA-internal/ioda-converters/pull/671

I'm going to close this issue since we've got a fix in place now.

**Comment by amfox37** [comment]
@srherbener 
I currently have about half of the ioda-converter tests failing with this error using a recent build of the jedi-stack on my MacBook Pro for the first time. I don't _think_ anything can be too wrong with the stack as all ioda tests are passing (and various model interfaces) Andy ideas I can try?

**Comment by srherbener** [comment]
@amfox37 I just tried building ioda-bundle with the converters included on my MacBook which is Big Sur, Clang 12.0.5, GNU Fortran 11.2.0. All the tests pass.

First thing is that you need to build the converters using ioda-bundle (which you are probably already doing). To do this use "-DBUILD_IODA_CONVERTERS=ON" on the ecbuild command line.

If you are already using ioda-bundle, then could you send me a sample from the error messages you are getting from the converters tests. Thanks!

**Comment by srherbener** [comment]
@amfox37 I've created a new issue in ZenHub (jcsda-internal/ioda/issues/581) to track this problem.
